### PR TITLE
Sms templates

### PR DESCRIPTION
**Pull request by 93Kamuran**
SMS settings page created.
The infrastructure of parsing parametric SMS sentences is created.

### CODE REVIEW

File: Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    protected $references;
+    protected $body = '';
+    protected $receiver;
+    protected $callback;
+
+    public function __construct($data, SmsBodyService $smsBodyService)
+    {
+        $this->smsBodyService = $smsBodyService;
+        $this->data = $data;
+        $this->validateReferences();
```
Comment (alchalade) [inline Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php:25]: Suggestion: move that logic to sendSms .

File: Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        $this->body .= $smsObject->parseSms($smsBody->body);
+    }
+
+    public function prepareFooter()
+    {
+        $smsBody = $this->smsBodyService->getSmsBodyByReference($this->references['footer']);
+        $this->body .= $smsBody->body;
+    }
+
+    private function validateReferences()
```
Comment (alchalade) [inline Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php:79]: you could use it to call the preparation methods if the key is in the references array

File: Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        //add sms to sms_gateway
+        resolve('SmsProvider')
+            ->sendSms(
+                $this->receiver,
+                $this->body,
+                $this->callback
+            );
+    }
+
+    public function prepareHeader()
```
Comment (alchalade) [inline Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php:56]: That part will break if the given reference is not stored in the database, error handling missing.

**Review by alchalade — CHANGES_REQUESTED** [review]

File: Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    public function sendSms()
+    {
+        if (config('app.debug')) {
+            Log::debug('Send sms on debug is not allowed in debug mode', ['number' => $this->receiver, 'message' => $this->body]);
+            return;
+        }
+       if (!($this->data instanceof Transaction) && !($this->data instanceof AssetRate)) {
+            $nullSmsBodies = $this->smsBodyService->getNullBodies();
+            if (count($nullSmsBodies)) {
+                Log::debug('Send sms rejected, some of sms bodies are null', ['Sms Bodies' => $nullSmsBodies]);
```
[the remark this answers is not in the document]
Comment (93Kamuran) [reply Website/htdocs/mpmanager/app/Sms/Senders/SmsSender.php:43]: Critical Log applied.

File: Website/htdocs/mpmanager/app/Http/Controllers/TransactionController.php [hunk 5]
```diff
@@ -805,54 +805,4 @@ private function determinePeriod($period): ?array
         return $comparisionPeriod;
     }
 
-
```
[the remark this answers is not in the document]
Comment (93Kamuran) [reply Website/htdocs/mpmanager/app/Http/Controllers/TransactionController.php:808]: I don't understand that.

File: Website/htdocs/mpmanager/app/Http/Controllers/SmsController.php [hunk 6]
```diff
@@ -50,26 +53,32 @@ class SmsController extends Controller
...
         ConnectionType $connectionType,
-        MeterParameter $meterParameter
+        MeterParameter $meterParameter,
+        SmsResendInformationKeyService $smsResendInformationKeyService
     ) {
         $this->sms = $sms;
         $this->person = $person;
         $this->connectionGroup = $connectionGroup;
         $this->connectionType = $connectionType;
         $this->meterParameter = $meterParameter;
```
[the remark this answers is not in the document]
Comment (93Kamuran) [reply Website/htdocs/mpmanager/app/Http/Controllers/SmsController.php:81]: Please check it again, I don't understand.

**Pull request merged**